src/io.rs

```diff
@@ -0,0 +1,502 @@
+                    e.encode(buffer)?;
+                }
+            }
+        };
+
+        Ok(())
+    }
+}
+
+impl<T: Deserialize> Deserialize for Vec<T> {
+    // Decode only the capacity of the vector. Elements will be decoded in the `decode_dynamic` method.
+    // The capacity is needed for iteration there.
+    fn decode_static<I: Input + ?Sized>(buffer: &mut I) -> Result<Self, Error> {
+        let cap: usize = usize::decode(buffer)?;
+
+        Ok(Vec::with_capacity(cap))
+    }
+
+    fn decode_dynamic<I: Input + ?Sized>(&mut self, buffer: &mut I) -> Result<(), Error> {
+        for _ in 0..self.capacity() {
+            // Bytes - Vec<u8> it a separate case without unpadding for each element.
+            // It should unpadded at the end if is not % ALIGN
+            match T::TYPE {
+                Type::U8 => {
+                    let byte = buffer.read_byte()?;
+                    // SAFETY: `Type::U8` implemented only for `u8`, so it is `Vec<u8>`.
+                    let _self =
+                        unsafe { ::core::mem::transmute::<&mut Vec<T>, &mut Vec<u8>>(self) };
+                    _self.push(byte);
+                }
+                // Spec doesn't say how to deserialize arrays with unaligned
+                // primitives(as `u16`, `u32`, `usize`), so unpad them.
+                _ => {
+                    self.push(T::decode(buffer)?);
+                }
+            };
+        }
+
+        if let Type::U8 = T::TYPE {
+            buffer.skip(fill_bytes(self.capacity()))?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<const N: usize, T: Serialize> Serialize for [T; N] {
+    fn encode_static<O: Output + ?Sized>(&self, buffer: &mut O) -> Result<(), Error> {
+        // Bytes - [u8; N] it a separate case without padding for each element.
+        // It should padded at the end if is not % ALIGN
+        match T::TYPE {
+            Type::U8 => {
+                // SAFETY: `Type::U8` implemented only for `u8`.
+                let bytes = unsafe { ::core::mem::transmute::<&[T; N], &[u8; N]>(self) };
+                buffer.write(bytes.as_slice())?;
+                for _ in 0..fill_bytes(N) {
+                    buffer.push_byte(0)?;
+                }
+            }
+            _ => {
+                for e in self.iter() {
+                    e.encode_static(buffer)?;
+                }
+            }
+        };
+
+        Ok(())
+    }
+
+    fn encode_dynamic<O: Output + ?Sized>(&self, buffer: &mut O) -> Result<(), Error> {
+        // All primitives have only static part, so skip dynamic encoding for them.
+        match T::TYPE {
+            Type::Unknown => {
+                for e in self.iter() {
+                    e.encode_dynamic(buffer)?;
+                }
+            }
+            _ => {}
+        };
+
+        Ok(())
+    }
+}
+
+impl<const N: usize, T: Deserialize> Deserialize for [T; N] {
+    fn decode_static<I: Input + ?Sized>(buffer: &mut I) -> Result<Self, Error> {
```

Comment:
It contains a lot of `unsafe`, but I did the same change here: https://github.com/paritytech/parity-scale-codec/pull/299

So it should be safe=)